sdk/dotnet/Pulumi/Deployment/GrpcEngine.cs

```diff
@@ -16,27 +16,36 @@ internal class GrpcEngine : IEngine
         private readonly Engine.EngineClient _engine;
         // Using a static dictionary to keep track of and re-use gRPC channels
         // According to the docs (https://docs.microsoft.com/en-us/aspnet/core/grpc/performance?view=aspnetcore-6.0#reuse-grpc-channels), creating GrpcChannels is expensive so we keep track of a bunch of them here
-        private static ConcurrentDictionary<string, GrpcChannel> _engineChannels = new ConcurrentDictionary<string, GrpcChannel>();
-
+        private static readonly ConcurrentDictionary<string, GrpcChannel> _engineChannels = new ConcurrentDictionary<string, GrpcChannel>();
+        private static readonly object _channelsLock = new object();
         public GrpcEngine(string engineAddress)
         {
+            // Allow for insecure HTTP/2 transport (only needed for netcoreapp3.x)
+            // https://docs.microsoft.com/en-us/aspnet/core/grpc/troubleshoot?view=aspnetcore-6.0#call-insecure-grpc-services-with-net-core-client
+            AppContext.SetSwitch("System.Net.Http.SocketsHttpHandler.Http2UnencryptedSupport", true);
             // maxRpcMessageSize raises the gRPC Max Message size from `4194304` (4mb) to `419430400` (400mb)
-            var maxRpcMessageSize = 400 * 1024 * 1024;
-            if (!_engineChannels.ContainsKey(engineAddress))
+            const int maxRpcMessageSize = 400 * 1024 * 1024;
+            if (_engineChannels.TryGetValue(engineAddress, out var engineChannel))
             {
-                // Allow for insecure HTTP/2 transport (only needed for netcoreapp3.x)
-                // https://docs.microsoft.com/en-us/aspnet/core/grpc/troubleshoot?view=aspnetcore-6.0#call-insecure-grpc-services-with-net-core-client
-                AppContext.SetSwitch("System.Net.Http.SocketsHttpHandler.Http2UnencryptedSupport", true);
-                // Inititialize the engine channel once for this address
-                _engineChannels[engineAddress] = GrpcChannel.ForAddress(new Uri($"http://{engineAddress}"), new GrpcChannelOptions
-                {
-                    MaxReceiveMessageSize = maxRpcMessageSize,
-                    MaxSendMessageSize = maxRpcMessageSize,
-                    Credentials = Grpc.Core.ChannelCredentials.Insecure,
-                });
+                // A channel already exists for this address
+                this._engine = new Engine.EngineClient(engineChannel);
             }
+            else 
+            {
+                lock (_channelsLock)
+                {
+                    // Inititialize the engine channel once for this address
```

Comment:
Between TryGetValue and taking the lock someone else could of added to the dictionary so we need another TryGetValue here.